biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts

```diff
@@ -282,11 +291,102 @@ export class Biogrid implements Grid {
     const rows = Math.ceil(numberOfGridItems / cols);
     const positions: ItemPosition[] = [];
     for (let i = 0; i < numberOfGridItems; i++) {
-      positions.push({
-        x: (((i % cols) + 0.5) / cols) * townSize.width,
-        y: ((Math.floor(i / cols) + 0.5) / rows) * townSize.height,
-      });
+      const newPositionUnverified = {
+        x: this.roundToGridDistance(
+          (((i % cols) + 0.5) / cols) * townSize.width
+        ),
+        y: this.roundToGridDistance(
+          ((Math.floor(i / cols) + 0.5) / rows) * townSize.height
+        ),
+      };
+      const newPosition = this.findNearestUnoccupiedPosition(
+        newPositionUnverified,
+        townSize
+      );
+      positions.push(newPosition);
+      this.itemInPosition[this.formatItemPosition(newPosition)] = true;
     }
     return positions;
   }
+
+  /**
+   * Find the nearest unoccupied position to {@code pos} by looking looking in a spiral with pos at its center
+   * First the space immediately right of pos is checked, then the one above, then to the left, then below, then two right spaces out, two right up, etc
+   */
+  private findNearestUnoccupiedPosition(
+    pos: ItemPosition,
+    townSize: TownSize
+  ): ItemPosition {
+    let radius = bioconstants.GRID_DISTANCES.INCREMENTS_KM;
+    let angle = 0;
+    let outOfBoundsCount = 0;
+    let xOffset = 0,
+      yOffset = 0;
+    let newPos = { x: pos.x + xOffset, y: pos.y + yOffset };
+    // If {@code outOfBoundsCount} is greater than 3, then that means the upwards, left, right, and down
+    // Are all out of bounds. Thus there is no where left to place the item
+    while (
+      (this.positionOutOfBounds(newPos, townSize) ||
+        this.positionOccupied(newPos)) &&
+      outOfBoundsCount < 4
+    ) {
+      if (this.positionOutOfBounds(newPos, townSize)) {
+        outOfBoundsCount++;
+      }
+      switch (angle) {
+        case 0:
+          yOffset = 0;
+          xOffset = radius;
+          break;
+        case 90:
+          xOffset = 0;
+          yOffset = radius;
+          break;
+        case 180:
+          xOffset = -1 * radius;
+          yOffset = 0;
+          break;
+        case 270:
+          xOffset = 0;
+          yOffset = -1 * radius;
+          break;
+      }
+      newPos = { x: pos.x + xOffset, y: pos.y + yOffset };
+      // Increment the angle by 90 degrees
+      angle = angle + 90;
+      if (angle === 360) {
+        radius += bioconstants.GRID_DISTANCES.INCREMENTS_KM;
+        // Reset the angle
+        angle = 0;
+      }
+    }
+    if (outOfBoundsCount > 3) {
```

Comment:
I do think it's a bit confusing having outOfBoundsCount being compared to 4 above and to 3 here. Can we make this consistent with `>3` and `<=3` above or the other way with 4?

Reply:
ok sure